### PR TITLE
rgw: fix PUT on versioned bucket fails with NoSuchKey

### PR DESCRIPTION
**Pull request by mkogan1**
Fixes: https://tracker.ceph.com/issues/48709

Signed-off-by: Mark Kogan <mkogan@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mkogan1** [comment]
submitted teuthology full rgw suite:
https://pulpito.ceph.com/mkogan-2020-12-23_18:51:19-rgw-wip-rgw-broken-pipe_i01-distro-basic-smithi/

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -6009,7 +6009,11 @@ int RGWRados::Bucket::UpdateIndex::guard_reshard(BucketShard **pbs, std::functio
   for (int i = 0; i < NUM_RESHARD_RETRIES; ++i) {
     int ret = get_bucket_shard(&bs);
     if (ret < 0) {
-      ldout(store->ctx(), 5) << "failed to get BucketShard object: ret=" << ret << dendl;
+      if (ret == -ENOENT) {
+        ldout(store->ctx(), 5) << "failed to get BucketShard object: ret=" << ret << ", modifying to " << -ERR_INTERNAL_ERROR << dendl;
+        ret = -ERR_INTERNAL_ERROR;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:6014]: change on this path seems sane--nothing has I think happened, so request seems definitely retryable

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -6091,7 +6095,11 @@ int RGWRados::Bucket::UpdateIndex::complete(int64_t poolid, uint64_t epoch,
 
   int ret = get_bucket_shard(&bs);
   if (ret < 0) {
-    ldout(store->ctx(), 5) << "failed to get BucketShard object: ret=" << ret << dendl;
+    if (ret == -ENOENT) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:6098]: less sure about this path;  what request are we executing, and what has happened in the current request to this point?

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -6640,7 +6648,11 @@ int RGWRados::guard_reshard(BucketShard *bs,
   for (int i = 0; i < NUM_RESHARD_RETRIES; ++i) {
     r = bs->init(pobj->bucket, *pobj, nullptr /* no RGWBucketInfo */);
     if (r < 0) {
-      ldout(cct, 5) << "bs.init() returned ret=" << r << dendl;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:6643]: assuming this is the same as RGWRados::Bucket::UpdateIndex::guard_reshard()--but I didn't realize there were two guard_reshard() methods

File: src/rgw/rgw_rados.cc [hunk 4]
```diff
@@ -6877,7 +6889,11 @@ int RGWRados::bucket_index_read_olh_log(const RGWBucketInfo& bucket_info, RGWObj
   int ret =
     bs.init(obj_instance.bucket, obj_instance, nullptr /* no RGWBucketInfo */);
   if (ret < 0) {
-    ldout(cct, 5) << "bs.init() returned ret=" << ret << dendl;
+    if (ret == -ENOENT) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:6892]: here again I'm unsure whether to trust that the current request has had no side effect/is safely retryable

File: src/rgw/rgw_rados.cc [hunk 5]
```diff
@@ -6982,7 +6998,11 @@ int RGWRados::bucket_index_trim_olh_log(const RGWBucketInfo& bucket_info, RGWObj
...
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.cc:7001]: ditto here;  what has happened to this point, for the current request?

**Comment by cbodley** [comment]
this looks valid as a workaround, but i think we're still missing something here

once a reshard completes, `guard_reshard()` calls `target->update_bucket_id()` so it can retry the op against the new bucket instance/index. we read that new bucket instance metadata into `RGWRados::Bucket::bucket_info` and use that for the retry

but when writing versioned objects, there are multiple calls to `guard_reshard()` in `RGWRados::Object::Write::_do_write_meta()`. if the first one hits this `update_bucket_id()` case, the updated `RGWRados::Bucket::bucket_info` will only be used for the retry - later calls to `guard_reshard()` would still be using the original pre-resharded version of the bucket_info, and so get ENOENT once reshard finishes deleting it

it sounds like everything under `_do_write_meta()` should be passing the bucket info as a mutable reference `RGWBucketInfo&`, so that the call to `update_bucket_id()` applies throughout. does that make sense?

**Comment by mattbenjamin** [comment]
@cbodley it sounds right in that uses after reshard starts would be uniform?

**Comment by mkogan1** [comment]
Unstale please

**Comment by mkogan1** [comment]
fixed by
1. https://github.com/ceph/ceph/pull/45345 -- cls/rgw: rgw_dir_suggest_changes detects race with completion
2. https://github.com/ceph/ceph/pull/45300 -- rgw: Update "CEPH_RGW_DIR_SUGGEST_LOG_OP" for remove entries

**Pull request closed without merging**